File: Whatsapp-OCR-CRM/backend/src/services/automation.service.ts
import { prisma } from "../lib/prisma";
import { Prisma } from "@prisma/client";
import { logger } from "../utils/logger";
import { OPEN_PIPELINE_STAGES } from "../config/automation-rules";
import {
  enqueueAutomationJob,
  recordCronTick,
} from "./automation-execution.service";

function daysAgo(days: number): Date {
  return new Date(Date.now() - days * 24 * 60 * 60 * 1000);
}

async function customerRepliedSince(customerId: string, since: Date): Promise<boolean> {
  const conv = await prisma.conversation.findFirst({
    where: { customerId },
    include: {
      messages: {
        where: { direction: "INBOUND", createdAt: { gt: since } },
        take: 1,
      },
    },
  });
  return (conv?.messages?.length ?? 0) > 0;
}

async function getLatestEnquiry(customerId: string) {
  return prisma.enquiry.findFirst({
    where: { customerId },
    orderBy: { createdAt: "desc" },
    include: { quotation: true },
  });
}

async function hasRecentExecution(ruleId: string, customerId: string, since: Date): Promise<boolean> {
  const existing = await prisma.scheduledJob.findFirst({
    where: {
      ruleId,
      customerId,
      status: { in: ["COMPLETED", "PENDING", "PROCESSING"] },
      isTest: false,
      createdAt: { gte: since },
    },
  });
  return !!existing;
}

export type TriggerResult = { queued: number; skipped: number };

export async function scheduleInquiryFollowup(
  ruleId: string,
  customerId: string,
  days: number,
  enquiryId: string,
  quotationNumber: string
): Promise<string> {
  const rule = await prisma.automationRule.findUnique({ where: { id: ruleId } });
  const customer = await prisma.customer.findUnique({ where: { id: customerId } });
  if (!rule || !customer) throw new Error("Rule or customer not found");

  const actionParams = rule.actionParams as Record<string, unknown>;
  const templateName = String(actionParams?.templateName || "mahabir_inquiry_followup");

  const scheduledAt = new Date(Date.now() + days * 24 * 60 * 60 * 1000);

  const scheduledJob = await prisma.scheduledJob.create({
    data: {
      ruleId,
      customerId,
      scheduledAt,
      status: "PENDING",
      metadata: { enquiryId, quotationNumber, templateName } as Prisma.InputJsonValue,
    },
  });

  const { automationQueue } = await import("../jobs/queues");
  const bullJob = await automationQueue.add(
    "inquiry_followup",
    {
      scheduledJobId: scheduledJob.id,
      ruleId,
      customerId,
      phone: customer.phone,
      templateName,
      variables: [customer.name || "Customer", quotationNumber],
      enquiryId,
      quotationNumber,
    },
    { delay: days * 24 * 60 * 60 * 1000 }
  );

  await prisma.scheduledJob.update({
    where: { id: scheduledJob.id },
    data: { bullJobId: String(bullJob.id) },
  });

  return scheduledJob.id;
}

export async function triggerPriceDrop(inventoryId: string, oldRate: number, newRate: number) {
  if (newRate >= oldRate) return { queued: 0, skipped: 0 };

  let queued = 0;
  let skipped = 0;

  try {
    const rules = await prisma.automationRule.findMany({
      where: { triggerType: "price_drop_alert", isActive: true },
    });
    if (rules.length === 0) return { queued, skipped };

    const items = await prisma.enquiryItem.findMany({
      where: { inventoryId },
      select: { enquiryId: true },
    });
    const enquiryIds = Array.from(new Set(items.map((it) => it.enquiryId)));
    if (enquiryIds.length === 0) return { queued, skipped };

    const enquiries = await prisma.enquiry.findMany({
      where: { id: { in: enquiryIds } },
      select: { customerId: true },
    });
    const customerIds = Array.from(new Set(enquiries.map((e) => e.customerId)));
    const inventory = await prisma.inventory.findUnique({ where: { id: inventoryId } });
    if (!inventory) return { queued, skipped };

    const dropPct = oldRate > 0 ? ((oldRate - newRate) / oldRate) * 100 : 0;

    for (const rule of rules) {
      const triggerParams = rule.triggerParams as Record<string, unknown>;
      const threshold = Number(triggerParams?.threshold ?? 0);
      const maxInquiryAgeDays = Number(triggerParams?.maxInquiryAgeDays ?? 30);
      const maxAgeCutoff = daysAgo(maxInquiryAgeDays);

      if (threshold > 0 && dropPct < threshold) continue;

      const actionParams = rule.actionParams as Record<string, unknown>;
      const templateName = String(actionParams?.templateName || "mahabir_price_drop");

      for (const customerId of customerIds) {
        const customer = await prisma.customer.findUnique({ where: { id: customerId } });
        if (!customer || customer.stage === "Closed") {
          skipped++;
          continue;
        }

        const latestEnquiry = await getLatestEnquiry(customerId);
        if (!latestEnquiry || latestEnquiry.createdAt < maxAgeCutoff) {
          skipped++;
          continue;
        }

        await enqueueAutomationJob("price_drop_alert", {
          ruleId: rule.id,
          customerId,
          phone: customer.phone,
          templateName,
          variables: [
            customer.name || "Customer",
            inventory.name,
            oldRate.toFixed(2),
            newRate.toFixed(2),
          ],
          inventoryId,
          oldRate,
          newRate,
        });
        queued++;
      }
    }
  } catch (error) {
    logger.error(`Failed to trigger price drop automation: ${error}`);
  }

  return { queued, skipped };
}

export async function triggerRepeatEngagement(): Promise<TriggerResult> {
  let queued = 0;
  let skipped = 0;

  try {
    const rules = await prisma.automationRule.findMany({
      where: { triggerType: "repeat_engagement", isActive: true },
    });

    for (const rule of rules) {
      const triggerParams = rule.triggerParams as Record<string, unknown>;
      const inactiveDays = Number(triggerParams?.inactiveDays ?? triggerParams?.days ?? 30);
      const stages = (triggerParams?.stages as string[]) || [...OPEN_PIPELINE_STAGES];
      const cutoff = daysAgo(inactiveDays);

      const actionParams = rule.actionParams as Record<string, unknown>;
      const templateName = String(actionParams?.templateName || "mahabir_repeat_engagement");

      const customers = await prisma.customer.findMany({
        where: { stage: { in: stages } },
      });

      let ruleQueued = 0;
      let ruleSkipped = 0;

      for (const customer of customers) {
        const latestEnquiry = await getLatestEnquiry(customer.id);
        if (!latestEnquiry || latestEnquiry.createdAt > cutoff) {
          ruleSkipped++;
          continue;
        }

        if (await hasRecentExecution(rule.id, customer.id, cutoff)) {
          ruleSkipped++;
          continue;
        }

        await enqueueAutomationJob("repeat_engagement", {
          ruleId: rule.id,
          customerId: customer.id,
          phone: customer.phone,
          templateName,
          variables: [customer.name || "Customer"],
        });
        ruleQueued++;
      }

      queued += ruleQueued;
      skipped += ruleSkipped;

      await recordCronTick({
        ruleId: rule.id,
        queued: ruleQueued,
        skipped: ruleSkipped,
        note:
          ruleQueued === 0
            ? `Daily run: no customers matched (${ruleSkipped} skipped — enquiry too recent or already contacted)`
            : `Daily run: ${ruleQueued} message(s) queued, ${ruleSkipped} skipped`,
      });
    }
  } catch (error) {
    logger.error(`Failed to trigger repeat engagement: ${error}`);
  }

  return { queued, skipped };
}

export async function triggerEnquiryReminder(): Promise<TriggerResult> {
  let queued = 0;
  let skipped = 0;

  try {
    const rules = await prisma.automationRule.findMany({
      where: { triggerType: "enquiry_reminder", isActive: true },
    });

    for (const rule of rules) {
      const triggerParams = rule.triggerParams as Record<string, unknown>;
      const daysSinceSent = Number(triggerParams?.daysSinceSent ?? triggerParams?.hours ?? 7);
      const sentCutoff = daysAgo(daysSinceSent);

      const actionParams = rule.actionParams as Record<string, unknown>;
      const templateName = String(actionParams?.templateName || "mahabir_enquiry_reminder");

      const customers = await prisma.customer.findMany({
        where: { stage: { not: "Closed" } },
      });

      let ruleQueued = 0;
      let ruleSkipped = 0;

      for (const customer of customers) {
        const lastSent = await prisma.enquiry.findFirst({
          where: {
            customerId: customer.id,
            status: "SENT",
            quotation: { is: { sentAt: { not: null } } },
          },
          orderBy: { updatedAt: "desc" },
          include: { quotation: true },
        });

        if (!lastSent?.quotation?.sentAt) {
          ruleSkipped++;
          continue;
        }
        if (lastSent.quotation.sentAt > sentCutoff) {
          ruleSkipped++;
          continue;
        }
        if (await customerRepliedSince(customer.id, lastSent.quotation.sentAt)) {
          ruleSkipped++;
          continue;
        }
        if (await hasRecentExecution(rule.id, customer.id, sentCutoff)) {
          ruleSkipped++;
          continue;
        }

        const quoteRef = lastSent.quotation.number || lastSent.id.slice(0, 8).toUpperCase();

        await enqueueAutomationJob("enquiry_reminder", {
          ruleId: rule.id,
          customerId: customer.id,
          enquiryId: lastSent.id,
          phone: customer.phone,
          templateName,
          variables: [customer.name || "Customer", quoteRef],
        });
        ruleQueued++;
      }

      queued += ruleQueued;
      skipped += ruleSkipped;

      await recordCronTick({
        ruleId: rule.id,
        queued: ruleQueued,
        skipped: ruleSkipped,
        note:
          ruleQueued === 0
            ? `Daily run: no sent quotations need reminder (${ruleSkipped} customers checked)`
            : `Daily run: ${ruleQueued} reminder(s) queued, ${ruleSkipped} skipped`,
      });
    }
  } catch (error) {
    logger.error(`Failed to trigger enquiry reminder: ${error}`);
  }

  return { queued, skipped };
}

export async function runCronRules() {
  const repeat = await triggerRepeatEngagement();
  const reminder = await triggerEnquiryReminder();
  return {
    repeatEngagement: repeat,
    enquiryReminder: reminder,
    totalQueued: repeat.queued + reminder.queued,
    totalSkipped: repeat.skipped + reminder.skipped,
  };
}

File: Whatsapp-OCR-CRM/backend/services/automation_service.py
"""Automation service.

Implements four trigger handlers and a simple scheduler that runs in the
background. We don't have BullMQ; we use asyncio + DB-backed ScheduledJob rows.
"""
import asyncio
import logging
from datetime import datetime, timezone, timedelta
from typing import Dict, Any, List

from database import db
from models import ScheduledJob
from services.whatsapp import send_template_message

log = logging.getLogger(__name__)


async def schedule_inactivity_followup(rule: Dict[str, Any], customer_id: str, days: int):
    """Schedule a follow-up N days after a quotation is sent."""
    when = (datetime.now(timezone.utc) + timedelta(days=days)).isoformat()
    job = ScheduledJob(
        ruleId=rule["id"], customerId=customer_id, scheduledAt=when, payload={"type": "inactivity_followup"}
    )
    await db.scheduled_jobs.insert_one(job.model_dump())
    return job.id


async def trigger_price_drop(inventory_id: str, old_rate: float, new_rate: float):
    """When inventory rate drops, find customers who enquired about it & notify."""
    if new_rate >= (old_rate or new_rate):
        return  # not a drop
    rules = await db.automation_rules.find({"triggerType": "price_drop_alert", "isActive": True}).to_list(20)
    if not rules:
        return

    # Find customers who enquired about this product
    items = await db.enquiry_items.find({"inventoryId": inventory_id}, {"_id": 0}).to_list(500)
    enquiry_ids = list({it["enquiryId"] for it in items})
    if not enquiry_ids:
        return
    enquiries = await db.enquiries.find({"id": {"$in": enquiry_ids}}, {"_id": 0}).to_list(500)
    customer_ids = list({e["customerId"] for e in enquiries})
    inv = await db.inventory.find_one({"id": inventory_id}, {"_id": 0})
    if not inv:
        return

    for rule in rules:
        template = rule.get("actionParams", {}).get("templateName", "price_drop_alert")
        for cid in customer_ids:
            cust = await db.customers.find_one({"id": cid}, {"_id": 0})
            if not cust:
                continue
            variables = [cust.get("name") or "Customer", inv["name"], f"{old_rate:.2f}", f"{new_rate:.2f}"]
            await send_template_message(
                conversation_id=None, phone=cust["phone"], template_name=template, variables=variables
            )
            # log a scheduled_jobs row for audit
            job = ScheduledJob(
                ruleId=rule["id"], customerId=cid,
                scheduledAt=datetime.now(timezone.utc).isoformat(),
                status="COMPLETED",
                payload={"type": "price_drop_alert", "inventoryId": inventory_id, "oldRate": old_rate, "newRate": new_rate},
            )
            await db.scheduled_jobs.insert_one(job.model_dump())


async def trigger_enquiry_reminder():
    """For DRAFT enquiries older than N hours, log an internal alert (no WhatsApp)."""
    rules = await db.automation_rules.find({"triggerType": "enquiry_reminder", "isActive": True}).to_list(20)
    for rule in rules:
        hours = int(rule.get("triggerParams", {}).get("hours", 24))
        cutoff = (datetime.now(timezone.utc) - timedelta(hours=hours)).isoformat()
        stale = await db.enquiries.find({"status": "DRAFT", "createdAt": {"$lt": cutoff}}, {"_id": 0}).to_list(200)
        for e in stale:
            # avoid duplicate alerts: only log once per enquiry per rule
            existing = await db.scheduled_jobs.find_one({"ruleId": rule["id"], "payload.enquiryId": e["id"]})
            if existing:
                continue
            job = ScheduledJob(
                ruleId=rule["id"], customerId=e["customerId"],
                scheduledAt=datetime.now(timezone.utc).isoformat(),
                status="COMPLETED",
                payload={"type": "enquiry_reminder", "enquiryId": e["id"]},
            )
            await db.scheduled_jobs.insert_one(job.model_dump())


async def trigger_repeat_engagement():
    """For customers whose last enquiry was > N days ago, send re-engagement template."""
    rules = await db.automation_rules.find({"triggerType": "repeat_engagement", "isActive": True}).to_list(20)
    for rule in rules:
        days = int(rule.get("triggerParams", {}).get("days", 30))
        cutoff = (datetime.now(timezone.utc) - timedelta(days=days)).isoformat()
        # find customers whose latest enquiry is older than cutoff
        # naive scan: list customers, find their max enquiry createdAt
        customers = await db.customers.find({}, {"_id": 0}).to_list(2000)
        for cust in customers:
            latest = await db.enquiries.find({"customerId": cust["id"]}, {"_id": 0}).sort("createdAt", -1).limit(1).to_list(1)
            if not latest:
                continue
            if latest[0]["createdAt"] > cutoff:
                continue
            # avoid sending more than once a month per rule per customer
            recent_job = await db.scheduled_jobs.find_one({
                "ruleId": rule["id"], "customerId": cust["id"],
                "scheduledAt": {"$gte": (datetime.now(timezone.utc) - timedelta(days=days)).isoformat()},
            })
            if recent_job:
                continue
            template = rule.get("actionParams", {}).get("templateName", "repeat_engagement")
            await send_template_message(conversation_id=None, phone=cust["phone"], template_name=template, variables=[cust.get("name") or "Customer"])
            job = ScheduledJob(
                ruleId=rule["id"], customerId=cust["id"],
                scheduledAt=datetime.now(timezone.utc).isoformat(),
                status="COMPLETED",
                payload={"type": "repeat_engagement"},
            )
            await db.scheduled_jobs.insert_one(job.model_dump())


async def run_due_jobs():
    """Process scheduled jobs whose time has come (inactivity_followup type)."""
    now_iso = datetime.now(timezone.utc).isoformat()
    due = await db.scheduled_jobs.find({"status": "PENDING", "scheduledAt": {"$lte": now_iso}}, {"_id": 0}).to_list(50)
    for job in due:
        await db.scheduled_jobs.update_one({"id": job["id"]}, {"$set": {"status": "PROCESSING"}})
        try:
            ptype = job.get("payload", {}).get("type")
            rule = await db.automation_rules.find_one({"id": job["ruleId"]}, {"_id": 0})
            cust = await db.customers.find_one({"id": job["customerId"]}, {"_id": 0})
            if not rule or not cust:
                await db.scheduled_jobs.update_one({"id": job["id"]}, {"$set": {"status": "FAILED"}})
                continue

            if ptype == "inactivity_followup":
                # Check no reply since quotation sent
                template = rule.get("actionParams", {}).get("templateName", "inactivity_followup")
                await send_template_message(conversation_id=None, phone=cust["phone"], template_name=template, variables=[cust.get("name") or "Customer"])
            await db.scheduled_jobs.update_one({"id": job["id"]}, {"$set": {"status": "COMPLETED"}})
        except Exception as e:
            log.exception("Job failed: %s", e)
            await db.scheduled_jobs.update_one({"id": job["id"]}, {"$set": {"status": "FAILED"}})


async def automation_loop():
    """Background loop. Runs every 30s in MOCK mode."""
    while True:
        try:
            await run_due_jobs()
            await trigger_enquiry_reminder()
        except Exception as e:
            log.exception("automation_loop tick failed: %s", e)
        await asyncio.sleep(30)
